File: FastForward Planner/src/edu/uno/ai/planning/ff/FastForwardSearch.java
package edu.uno.ai.planning.ff;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import edu.uno.ai.planning.Plan;
import edu.uno.ai.planning.Planner;
import edu.uno.ai.planning.Problem;
import edu.uno.ai.planning.Search;
import edu.uno.ai.planning.SearchLimitReachedException;
import edu.uno.ai.planning.ss.StateSpaceNode;
import edu.uno.ai.planning.ss.StateSpaceProblem;
import edu.uno.ai.planning.ss.StateSpaceSearch;

public class FastForwardSearch extends StateSpaceSearch {
	
	private int visited   = 0;
	private int expanded  = 0;
	private int nodeLimit = Planner.NO_NODE_LIMIT;

	public FastForwardSearch(StateSpaceProblem problem) {
		super(problem);
	}

	@Override
	public int countVisited() {
		return visited;
	}

	@Override
	public int countExpanded() {
		return expanded;
	}

	@Override
	public void setNodeLimit(int limit) {
		nodeLimit = limit;
		
	}

	@Override
	public Plan findNextSolution() {
		StateSpaceNode bestNode = root;
		int hValue = Integer.MAX_VALUE;
		int nextValue;
		List<StateSpaceNode> nextLevel = new ArrayList<StateSpaceNode>(1);
		nextLevel.add(root);		

		List<StateSpaceNode> helpful   = new LinkedList<StateSpaceNode>();
		List<StateSpaceNode> unhelpful = new LinkedList<StateSpaceNode>();
		
		boolean foundBetter;
		while (!nextLevel.isEmpty()) {
			foundBetter = false;
			helpful.clear();
			unhelpful.clear();
			//separate helpful states from unhelpful states
			for (StateSpaceNode stateNode : nextLevel) {
				Util.separateHelpful(problem, stateNode, helpful, unhelpful);
			}
			//evaluate helpful actions
			for (StateSpaceNode stateNode : helpful) {
				nextValue = getHeuristicValue(stateNode);
				if (problem.isSolution(stateNode.plan)) {
					return stateNode.plan;
				} else {
					if (nextValue < hValue) {
						hValue = nextValue;
						bestNode = stateNode;
						foundBetter = true;
					}
				}
			}
			//if helpful actions yield nothing, weighted A* search on everything
			if (!foundBetter) {
				for (StateSpaceNode stateNode : unhelpful) {
					nextValue = getHeuristicValue(stateNode);
					if (problem.isSolution(stateNode.plan)) {
						return stateNode.plan;
					} else {
						if (nextValue < hValue) {
							hValue = nextValue;
							bestNode = stateNode;
							foundBetter = true;
						}
					}
				}
			}
			//build next layer, enforcing breadth-first search if no better node was found
			if (!foundBetter) {
				nextLevel = makeNextLevel(helpful,unhelpful);
			} else {
				nextLevel = makeNextLevel(bestNode);
			}
			
		}
		return null;
	}
	
	private int getHeuristicValue(StateSpaceNode stateNode) {
		visited++;
		if (visited >= nodeLimit) {
			throw new SearchLimitReachedException();
		}
		return new FastForwardHeuristic(problem).hValue(stateNode.state);
	}

	private List<StateSpaceNode> makeNextLevel (StateSpaceNode parent) {
		List<StateSpaceNode> nextLevel = new LinkedList<StateSpaceNode>();
		parent.expand();
		expanded++;
		for (StateSpaceNode child : parent.children) {
			nextLevel.add(child);
		}
		return nextLevel;
	}
	
	private List<StateSpaceNode> makeNextLevel (List<StateSpaceNode>... lists) {
		List<StateSpaceNode> nextLevel = new LinkedList<StateSpaceNode>();
		for (List<StateSpaceNode> list : lists) {
			for (StateSpaceNode parent : list) {
				parent.expand();
				expanded++;
				for (StateSpaceNode child : parent.children) {
					nextLevel.add(child);
				}
			}
		}
		return nextLevel;
	}
}

File: UnitySokoban/Assets/Scripts/Planning/FastForward/FastForwardSearch.cs
using StateSpaceSearchProject;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Planning;

namespace FastForward
{
    public class FastForwardSearch : StateSpaceSearchET
    {

        private int visited = 0;
        private int expanded = 0;
        private int nodeLimit = Planner<Search>.NO_NODE_LIMIT;

        public FastForwardSearch(StateSpaceProblem problem)
            : base(problem)
        {
        }

        public override int countVisited()
        {
            return visited;
        }


        public override int countExpanded()
        {
            return expanded;
        }

        public override void setNodeLimit(int limit)
        {
            nodeLimit = limit;

        }

        public override Plan findNextSolution()
        {
            StateSpaceNode bestNode = root;
            int hValue = int.MaxValue;
            int nextValue;
            List<StateSpaceNode> nextLevel = new List<StateSpaceNode>(1);
            nextLevel.Add(root);

            List<StateSpaceNode> helpful = new List<StateSpaceNode>();
            List<StateSpaceNode> unhelpful = new List<StateSpaceNode>();

            bool foundBetter;
            while (nextLevel.Count != 0)
            {
                foundBetter = false;
                helpful.Clear();
                unhelpful.Clear();
                //separate helpful states from unhelpful states
                foreach (StateSpaceNode stateNode in nextLevel)
                {
                    Util.separateHelpful(problem, stateNode, helpful, unhelpful);
                }
                //evaluate helpful actions
                foreach (StateSpaceNode stateNode in helpful)
                {
                    if (ShutThisSuckaDown)
                    {
                        ShutThisSuckaDown = false;
                        return null;
                    }
                    nextValue = getHeuristicValue(stateNode);
                    if (problem.isSolution(stateNode.plan))
                    {
                        return stateNode.plan;
                    }
                    else
                    {
                        if (nextValue < hValue)
                        {
                            hValue = nextValue;
                            bestNode = stateNode;
                            foundBetter = true;
                        }
                    }
                }
                //if helpful actions yield nothing, weighted A* search on everything
                if (!foundBetter)
                {
                    foreach (StateSpaceNode stateNode in unhelpful)
                    {
                        if (ShutThisSuckaDown)
                        {
                            ShutThisSuckaDown = false;
                            return null;
                        }
                        nextValue = getHeuristicValue(stateNode);
                        if (problem.isSolution(stateNode.plan))
                        {
                            return stateNode.plan;
                        }
                        else
                        {
                            if (nextValue < hValue)
                            {
                                hValue = nextValue;
                                bestNode = stateNode;
                                foundBetter = true;
                            }
                        }
                    }
                }
                //build next layer, enforcing breadth-first search if no better node was found
                if (!foundBetter)
                {
                    nextLevel = makeNextLevel(helpful, unhelpful);
                }
                else
                {
                    nextLevel = makeNextLevel(bestNode);
                }

            }
            return null;
        }

        protected int getHeuristicValue(StateSpaceNode stateNode)
        {
            _currentNode = stateNode;
            visited++;
            if (visited >= nodeLimit && nodeLimit != -1)
            {
                throw new TimeoutException();
            }
            return new FastForwardHeuristic((StateSpaceProblem)problem).hValue(stateNode.state);
        }

        protected override int GetCost(StateSpaceNode node)
        {
            return new FastForwardHeuristic((StateSpaceProblem)problem).hValue(node.state);
        }

        protected List<StateSpaceNode> makeNextLevel(StateSpaceNode parent)
        {
            List<StateSpaceNode> nextLevel = new List<StateSpaceNode>();
            parent.expand();
            expanded++;
            foreach (StateSpaceNode child in parent.children)
            {
                nextLevel.Add(child);
            }
            return nextLevel;
        }

        protected List<StateSpaceNode> makeNextLevel(params List<StateSpaceNode>[] lists)
        {
            List<StateSpaceNode> nextLevel = new List<StateSpaceNode>();
            foreach (List<StateSpaceNode> list in lists)
            {
                foreach (StateSpaceNode parent in list)
                {
                    parent.expand();
                    expanded++;
                    foreach (StateSpaceNode child in parent.children)
                    {
                        nextLevel.Add(child);
                    }
                }
            }
            return nextLevel;
        }
    }
}
